File: pascal/fatorial.pas
program Fatorial;
        function fat(n: integer): integer;
        var
           i: integer;
        begin
             for i:=n-1 downto 1 do
                 n := n * i;
             fat := n;
        end;

        function fibonacci(n: integer): integer;
        var
           i: integer;
        begin
             if (n = 0) or (n = 1) then
                 fibonacci := 1
             else
                 fibonacci := fibonacci(n-2) + fibonacci(n-1);
        end;

        procedure imprimeValor (n: integer; function f(x:integer):integer);
        begin
             writeln('O valor da funcao aplicado a ', n, ' eh: ', f(n));
        end;

begin
     imprimeValor (6, fibonacci);
end.

File: haskell/fatorial.hs
-- fatorial x = if (x == 0) then 1 else x * fat(x - 1)

-- fatorial x
--     | x == 0 = 1
--     | otherwise = x * fat (x - 1)

fatorial 0 = 1
fatorial n = fatorial(n-1) * n

fib 1 = 1
fib 2 = 1
fib n = fib (n - 1) + fib (n - 2)

